src/Microsoft.AspNetCore.Sockets.Common/TextMessageBatchFormatter.cs

```diff
@@ -0,0 +1,239 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Sockets
+{
+    public static class TextMessageBatchFormatter
+    {
+        /// <summary>
+        /// Decodes the messages encoded in the provided data.
+        /// </summary>
+        /// <param name="data">The data containing the encoded messages to read</param>
+        /// <exception cref="FormatException">The input does not adhere to the protocol</exception>
+        /// <returns>A list of <see cref="Message"/> values decoded from the buffer</returns>
+        public static IEnumerable<Message> ReadMessages(ReadOnlySpan<byte> data)
+        {
+            if (data.Length < 1 || data[0] != 'T')
+            {
+                throw new FormatException("Missing 'T' prefix in Text Message Batch.");
+            }
+
+            int cursor = 1;
+            while (cursor < data.Length)
+            {
+                // Parse the length
+                int length = ParseLength(data, ref cursor);
+                cursor++;
+
+                // Parse the type
+                var type = ParseType(data, ref cursor);
+                cursor++;
+
+                // Read the payload
+                var payload = ParsePayload(data, length, type, ref cursor);
+                cursor++;
+
+                yield return new Message(payload, type, endOfMessage: true);
+            }
+        }
+
+        /// <summary>
+        /// Attempts to write the specified messages to the target buffer
+        /// </summary>
+        /// <param name="target">The buffer in which to encode the messages</param>
+        /// <param name="messages">The messages to encode</param>
+        /// <returns>A boolean indicating if there was enough space in the buffer to encode the message.</returns>
+        public static bool WriteMessages(Span<byte> target, IReadOnlyList<Message> messages)
```

Comment:
Also, should it not return how many bytes it has written? In general, why not follow signature pattern similar to primitive formatters: https://github.com/dotnet/corefxlab/blob/master/src/System.Text.Primitives/System/Text/Formatting/PrimitiveFormatter.cs